(7.0) Robotest custom image backports

wadells opened the pull request:
## Description
This is the 7.0 backport of custom robotest images and the related build changes.  This provides better validation of both tele build, and installing a cluster image with an application/hooks running inside of it.

If you've already reviewed the corresponding pull requests against master (see "Linked tickets and other PRs" below), I'd recommend you review the 7.0 specific changes, found after the first commit:

https://github.com/gravitational/gravity/pull/2224/files/16cff8895ce495db4b5593d87909755c381a12e4..3bc375523b83e0516c3229e62264663b4dfe739f

## Type of change
* New feature (non-breaking change which adds functionality)
* Internal change (not necessarily a bug fix or a new feature)

## Linked tickets and other PRs
* Contributes to: #1908, #1915
* Depends on https://github.com/gravitational/gravity.e/pull/4344
* Ports: #1967, #2133, #2150, #2197, #2198, #2212, #2223, #2233

## TODOs
- [x] Fix `--upgrade-via` testing
- [x] Self-review the change
- [x] Perform manual testing
- [ ] Address review feedback

## Implementation
I chose to move the construction of the intermediate upgrade tarball into the robotest Makefile, the general gravity Makefile no longer offers a target for this.  Furthermore, the intermediate upgrade tarball is distinct from the tarball used for general installs & upgrades, to test both different paths.

@a-palchikov these changes will need to be forward ported to master/7.1+ when `--upgrade-via` is forward ported.

## Testing done
The PR build is sufficient for automated runs.

The new upgrade-via testing can be seen here: https://console.cloud.google.com/logs/query;query=resource.type%3D%22project%22%0Alabels.__uuid__%3D%226a80f6c5-a07a-4a50-911d-ebddad942a49%22%0Alabels.__suite__%3D%22fbece971-9ae7-423d-81e9-a89729af16a9%22%0Aseverity%3E%3DINFO;timeRange=2020-10-13T18:00:00.000Z%2F2020-10-13T19:15:00.000Z?authuser=0&project=kubeadm-167321

Here is a run of the new make target:

<details><summary><code>make image-robotest-intermediate</code></summary>

```
walt@gcp:~/git/gravity$ make -C e/assets/robotest 
make: Entering directory '/home/walt/git/gravity/e/assets/robotest'
Usage: make [options] [target] ...

Options: run `make --help` for options

Targets:
  clean                            Remove build artifacts.
  clean-caches                     Remove cached tele binaries and gravity packages.
  docker-image-tele-build          Build a docker image suitable to run `tele build`.
  download-teles                   Download tele binaries needed to build robotest images.
  get-config                       Show robotest configuration.
  get-upgrade-versions             Show Gravity versions used in upgrade testing.
  help                             Show this message.
  image-opscenter                  Move opscenter.tar cluster image into place for test execution.
  image-robotest                   Build robotest cluster image using the current tele.
  image-robotest-intermediate      Build robotest cluster image with a prior LTS baked in through --upgrade-via.
  image-robotest-upgrade           Build robotest cluster images needed for upgrade testing.
  image-telekube                   Move telekube.tar cluster image into place for test execution.
  images                           Build all images necessary to run robotest CI tests.
  populate-caches                  Populate caches with tele binaries and gravity packages.
  run                              Run robotest.
make: Leaving directory '/home/walt/git/gravity/e/assets/robotest'

walt@gcp:~/git/gravity$ make -C e/assets/robotest GRAVITY_BUILDDIR=$(pwd)/e/build/current image-robotest-intermediate 2>&1 | grep -v '^time='
make: Entering directory '/home/walt/git/gravity/e/assets/robotest'
/home/walt/git/gravity/assets/robotest/tele_build.sh
+ docker run --rm=true --net=host --user=1000:1000 --group-add=998 -v /home/walt/git/gravity/e/build/current/tele:/home/walt/git/gravity/e/build/current/tele:ro -v /home/walt/git/gravity/assets/robotest/current:/home/walt/git/gravity/assets/robotest/current:ro -v /home/walt/git/gravity/e/build/current/packages:/home/walt/git/gravity/e/build/current/packages -v /home/walt/git/gravity/e/build/current/robotest/tmp/tmp.OmVBys8qRO:/home/walt/git/gravity/e/build/current/robotest/tmp/tmp.OmVBys8qRO -v /var/run/docker.sock:/var/run/docker.sock --tmpfs /tmp -w /home/walt/git/gravity/e/build/current/robotest/tmp/tmp.OmVBys8qRO debian-grande:buster-20201001-tele-build dumb-init /home/walt/git/gravity/e/build/current/tele build --upgrade-via=6.1.43 /home/walt/git/gravity/assets/robotest/current/app.yaml --state-dir=/home/walt/git/gravity/e/build/current/packages --version=7.0.21-dev.6 --output=/home/walt/git/gravity/e/build/current/robotest/tmp/tmp.OmVBys8qRO/robotest-7.0.21-dev.6-via-6.1.43.tar
Tue Oct 13 23:06:27 UTC Building cluster image robotest 7.0.21-dev.6
Tue Oct 13 23:06:27 UTC Selecting base image version
        Will use base image version 7.0.21-dev.4
Tue Oct 13 23:06:27 UTC Downloading dependencies from https://get.gravitational.io
Tue Oct 13 23:06:27 UTC Syncing packages for 6.1.43
Tue Oct 13 23:06:28 UTC Local package cache is up-to-date for 6.1.43
Tue Oct 13 23:06:28 UTC Local package cache is up-to-date for robotest:0.0.0
Tue Oct 13 23:06:28 UTC Embedding application container images
        Using local image quay.io/gravitational/debian-tall:buster
        Using local image quay.io/gravitational/debian-tall:buster
        Using local image k8s.gcr.io/echoserver:1.10
        Vendored image gravitational/debian-tall:buster
        Vendored image k8s.gcr.io/echoserver:1.10
Tue Oct 13 23:06:32 UTC Creating application
Tue Oct 13 23:06:35 UTC Generating the cluster image
        Still generating the cluster image (10 seconds elapsed)
Tue Oct 13 23:06:47 UTC Saving the image as /home/walt/git/gravity/e/build/current/robotest/tmp/tmp.OmVBys8qRO/robotest-7.0.21-dev.6-via-6.1.43.tar
Tue Oct 13 23:06:56 UTC Build finished in 29 seconds 
+ mv /home/walt/git/gravity/e/build/current/robotest/tmp/tmp.OmVBys8qRO/robotest-7.0.21-dev.6-via-6.1.43.tar /home/walt/git/gravity/e/build/current/robotest/images/robotest-7.0.21-dev.6-via-6.1.43.tar
make: Leaving directory '/home/walt/git/gravity/e/assets/robotest'
```
</details>

And a list of configs for 7.0:

<details><summary><code>make get-config</code></summary>

```
walt@gcp:~/git/gravity$ make -C e/assets/robotest get-config                                                                                 
make: Entering directory '/home/walt/git/gravity/e/assets/robotest'
install={"installer_url":"/images/telekube-7.0.21-dev.6.tar","nodes":3,"flavor":"three","role":"node","os":"ubuntu:18"}
install={"installer_url":"/images/opscenter-7.0.21-dev.6.tar","nodes":1,"flavor":"standalone","role":"node","os":"ubuntu:18","ops_advertise_addr":"example.com:443"}
install={"installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","storage_driver":"overlay2"}
resize={"installer_url":"/images/robotest-7.0.21-dev.6.tar","nodes":1,"to":3,"flavor":"one","role":"node","state_dir":"/var/lib/telekube","os":"ubuntu:18","storage_driver":"overlay2"}
shrink={"installer_url":"/images/robotest-7.0.21-dev.6.tar","nodes":3,"flavor":"three","role":"node","os":"redhat:7"}
upgrade={"from":"/images/robotest-7.0.7.tar","installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:16","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.20.tar","installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-6.3.18.tar","installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.13.tar","installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-7.0.12.tar","installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"ubuntu:18","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-6.1.43.tar","installer_url":"/images/robotest-7.0.21-dev.6.tar","flavor":"three","nodes":3,"role":"node","os":"redhat:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
upgrade={"from":"/images/robotest-5.5.54.tar","installer_url":"/images/robotest-7.0.21-dev.6-via-6.1.43.tar","flavor":"three","nodes":3,"role":"node","os":"centos:7","service_uid":997,"service_gid":994,"storage_driver":"overlay2"}
make: Leaving directory '/home/walt/git/gravity/e/assets/robotest
```
</details>

wadells committed:
Backport robotest images changes from master.

This is a verbatim change from the following list of master commits.
Adjustments for 7.0.x will be applied in a later commit.

(cherry picked from commit 654058bbaf5e20c68e288a247074dadb27836769)
(cherry picked from commit b03882d1b60f8a26ddf659f7be7332f936c4f3a6)
(cherry picked from commit 74c16169b780e6cb94d23c0ff0b87d6488f469b6)
(cherry picked from commit 22e2b85f1788eb69733f4be732decb025240c5f0)
(cherry picked from commit 23373a08bd1944def6c33e1281a9067924f483a5)
(cherry picked from commit caba0a2d9a3a5363f77fe74582dd52e6c8b9898d)
(cherry picked from commit 918782f3b0f4c1d1a22ffc18bc3a1aa1549413a0)
(cherry picked from commit 5766247b18081aabd44082c58ef6eb6385ec2101)

wadells committed:
Adjust pr and nightly config to version/7.0.x

This rebases the local changes in the branch on top of the refactored
robotest config.

squash 7.0.x

wadells committed:
Fix intermediate upgrade testing.

Unlike master, 7.0 tests upgrades through the `tele build --upgrade-via`
flag.  This changeset adds that flag to robotest builds, and removes it
from the general gravity build.

wadells committed:
Remove nightly config.

This is not currently used on 7.0.x and presents a maintenance burden.
Now that robotest make targets have been unified, removing this file
will not break anything.

a-palchikov commented on assets/robotest/download_tele.sh:
```diff
@@ -0,0 +1,26 @@
+TMP=$(mktemp -d -p "$BUILD_TMP")
+trap "rm -rf \"$TMP\"" exit
+
+wget --no-verbose "https://get.gravitational.com/gravity-$VERSION-linux-x86_64-bin.tar.gz" --directory-prefix "$TMP"
```
isn't this supposed to use s3 for oss content?

wadells replied:
Is it?  I'm using the bits we provide via the downloads page:

https://gravitational.com/gravity/download

If we want our tele downloads to come from S3, we should change it there first -- my interest is in testing what end users will use.

wadells commented on Jenkinsfile:
```diff
@@ -7,92 +7,111 @@ def propagateParamsToEnv() {
+    // offer a superset of PR/nightly parameters, and the extra ones will be
+    // lost when setRobotestParameters() is called -- 2020-04 walt
+    echo "Jenkins Job Parameters:"
       for (param in params) { echo "${param}" }
```
This will be fixed when I pull in the backports from #2223.  I need to complete that PR first.

wadells replied:
Fixed in e319172.

wadells committed:
Remove upgrade testing from 6.1.0.

6.1.0 cannot successfully install an application that uses status hooks.
The hook init container reliably crashes.

wadells committed:
Merge branch 'version/7.0.x' into walt/7.0-robotest-backports

wadells commented:
Build 11 CI failures:
* 2x https://github.com/gravitational/robotest/issues/270 (unrelated to the changes here)
* 3x 7.0.20 -> 7.0.21-dev (this version) failures -- resulting in the overall failure.  Still investigating these.

a-palchikov commented on assets/robotest/run.sh:
```diff
@@ -0,0 +1,118 @@
+	ABORT=
+	for v in $@ ; do
+		if [ ! -f "${v}" ] ; then
+			echo "${v} does not exist"
```
nit: output errors to stderr? `>&2 echo ...`

wadells replied:
I'm conflicted on this: it is a small change, but would also cause 7.0.x to unnecessarily diverge from master, resulting in a chance of future merge conflicts.  I'll pick it up in master first, and then backport it in a later bulk move.

a-palchikov reviewed (approved):
lgtm
